### PR TITLE
Files are dumped in stream to save memory

### PR DESCRIPTION
**Pull request by matllubos**

### CODE REVIEW

File: security/management/commands/purge_logs.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            return generator()
+
+        # according to the comment below
+        def __len__(self):
+            return 1
+
+    return StreamList()
+
+
+class Command(BaseCommand):
```
Comment (matllubos) [inline security/management/commands/purge_logs.py]: Parent is not required

File: security/management/commands/purge_logs.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    class StreamList(list):
+        def __iter__(self):
+            return generator()
+
+        # according to the comment below
+        def __len__(self):
+            return 1
+
+    return StreamList()
```
Comment (matllubos) [inline security/management/commands/purge_logs.py]: little bit hack, json.dump cannot serialize iterator WTF

File: security/management/commands/purge_logs.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+
+def lazy_serialize_qs_without_pk(qs):
+    def generator():
+        for obj in qs.iterator():
+            data = serializers.serialize('python', [obj])[0]
+
+            del data['pk']
+            yield data
```
Comment (matllubos) [inline security/management/commands/purge_logs.py]: I need it as iterable to save memory

File: security/management/commands/purge_logs.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+                if storage.exists('{}.json.gz'.format(log_file_name)):
+                    i = 1
+                    while storage.exists('{}({}).json.gz'.format(log_file_name, i)):
+                        i += 1
+                    log_file_name = '{}({})'.format(log_file_name, i)
+
+                self.stdout.write(4 * ' ' + 'generate backup file: {}.json.gz'.format(log_file_name))
+
+                with storage.open('{}.json.gz'.format(log_file_name), 'wb') as f:
```
Comment (matllubos) [inline security/management/commands/purge_logs.py]: it is tested on S3 and it works (everything is stream)

File: security/management/commands/purge_logs.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                    with TextIOWrapper(gzip.GzipFile(filename='{}.json'.format(log_file_name),
```
Comment (matllubos) [inline security/management/commands/purge_logs.py]: GzipFile requires bytes but json.dump string

File: security/management/commands/purge_logs.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                    self.backup_logs_and_clean_data(qs, options.get('backup'))
+                except IOError as ex:
+                    self.stderr.write(force_text(ex))
+
+    def get_model(self, type):
+        model = self.models.get(type)
+        if model:
+            return model
+        else:
+            raise CommandError('Type can be only input-request, output-request, command or celery')
```
Comment (rubickcz) [inline security/management/commands/purge_logs.py]: I would print contents of `self.models` here, hardcoding models here is duplication.

File: security/management/commands/purge_logs.py (hunk 4 above)
Comment (rubickcz) [reply security/management/commands/purge_logs.py]: It is stream by default?

Comment (matllubos) [reply security/management/commands/purge_logs.py]: yes

Comment (matllubos) [reply security/management/commands/purge_logs.py]: file is stream by default

File: security/management/commands/purge_logs.py (hunk 6 above)
Comment (matllubos) [reply security/management/commands/purge_logs.py]: yes good idea

File: security/management/commands/purge_logs.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
+import os
+import json
+import gzip
+
+from datetime import datetime, time, timedelta
+
+from six.moves import input
```
Comment (radimsuckr) [inline security/management/commands/purge_logs.py]: Do we really still want six?

File: security/management/commands/purge_logs.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        try:
+            amount = int(amount)
+        except ValueError:
+            raise CommandError('Invalid expiration format')
+
+        if unit not in UNIT_OPTIONS:
+            raise CommandError('Invalid expiration format')
+
+        model = self.get_model(type)
+        qs = model.objects.filter(**{'created_at__lte': UNIT_OPTIONS[unit](amount)})
```
Comment (radimsuckr) [inline security/management/commands/purge_logs.py]: Unnecessary dict comprehension. :)

File: security/management/commands/purge_logs.py (hunk 7 above)
Comment (matllubos) [reply security/management/commands/purge_logs.py]: no :)

File: security/management/commands/purge_logs.py (hunk 8 above)
Comment (matllubos) [reply security/management/commands/purge_logs.py]: yes it was refactorization. I will fix it

File: security/management/commands/purge_logs.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            if options.get('interactive'):
+                confirm = input('''
+                   You have requested a database reset.
+                   This will IRREVERSIBLY DESTROY any
+                   logs created before {} {}
+                   ago. That is a total of {} logs.
+                   Are you sure you want to do this?
+                   Type 'yes' to continue, or 'no' to cancel: '''.format(amount, unit, qs.count()))
+            else:
+                confirm = 'yes'
```
Comment (Formulka) [inline security/management/commands/purge_logs.py]: a constant would be better, even though it is a part of a string

File: security/management/commands/purge_logs.py (hunk 2 above)
Comment (Formulka) [reply security/management/commands/purge_logs.py]: how the hell did you even come up with this

File: security/management/commands/purge_logs.py (hunk 9 above)
Comment (matllubos) [reply security/management/commands/purge_logs.py]: maybe next time.

File: security/management/commands/purge_logs.py (hunk 2 above)
Comment (matllubos) [reply security/management/commands/purge_logs.py]: :D

**Pull request merged**